### ISSUE TITLE
Validate integer value parameters

### ISSUE THREAD

**Issue by kamenov** [kamenov/PDF-Table#12]
Make sure that values expected as int are int before they are used. i.e.

my $line_w = defined $arg{'border'} ? $arg{'border'} : 1 ;

IF $arg{'border'}  equals 'blabla' THEN $line_w = 'blabla' AND error will be thrown upon first use.

**Comment by PhilterPaper** [comment]
Yeah, this sort of validation for settings values should probably be done, especially for data type. I'll throw it in the hopper for an upcoming release.

**Comment by PhilterPaper** [comment]
Something like:

- for globals (not determined by `find_value()` call), a `validate_field(value, type, range)` call
- for settings that go through `find_value()`, add parms to pass to `validate_field()`

The object would be to flag a problem setting as early as possible, before it possibly gets used in determining default/fallback for other settings. Perhaps the parameters could include a fallback value if an invalid one was given? It would still have to output an error message, but at least PDF::Table could run with reasonable results and no system errors. It could be tricky to do compound settings such as 'h_rule_w' falling back to 'rule_w', where 'rule_w' is found to be bogus. To minimize annoyance, each bad setting should probably be limited to one error message, even though it may be used many times (such as once each cell in `find_value()`). However, if 2 different settings (e.g., `'rule_w' => -1` in one place and `'rule_w' => 400` in another) have different bogus values, they should probably both be flagged.

This doesn't have to be limited to integers. It could check for valid floats and strings too. The allowed 'range' might include minimum and maximum values, minimum only (e.g., 0 for border and rule widths), multiple sub-ranges, and even enumerated values. For strings, it would probably have to be an enumerated list of valid values.

**Comment by PhilterPaper** [comment]
Here is a possibility. Call a validation routine after deprecated names are modified to supported names, and before any settings are actually used. Go through global, row_props, column_props, cell_props like `find_value` does. An array of validation settings would be defined:
```
%validation = (
  {  # one entry, for rules and borders
     names => [ 'border_w', 'h_border_w', 'v_border_w', 'rule_w', 'h_rule_w', 'v_rule_w' ],
     type => number | int | float | string | bool | color ,  # here number
     # optional settings:
       min => minimum value, NOT included,
       Min => minimum value, inclusive,      # here, 0
       max => maximum value, NOT included,   # perhaps something like 100 reasonable for most applications?
       Max => maximum value, inclusive,
       enum => [ 'list', 'of', 'permitted', 'exact', 'values' ],
       multi => [ {min/Min=>, max/Max=>}, {min/Min=>, max/Max=>},... ],  # multiple ranges
       replace => value to use if invalid setting given (still give error message)
       default => default value if no setting given
  },
  { # some other setting  }, ...
              );
```
`bool` would get special treatment, permitting only 0/N/No/Off/F/False (converted to 0) and 1/Y/Yes/On/T/True (converted to 1), case-insensitive. min/max/enum/multi entries would be ignored for boolean.

A `string` could have only `enum` for a list of permitted strings.

This might be overly elaborate, but it probably covers the waterfront.  `int` is necessary, and `number` could allow either an int or a float (clearer than forcing a `float` to be used?). Only one of 'min' or 'Min' could be used at a time, and likewise for 'max'/'Max'.

A `color` type would check for a legitimate numeric/hex code (a number of formats), and might check names against a standard list. An `enum` list might be added to further restrict allowed values.

I think all settings are currently scalar values. If any lists are added, size limits would have to be added.

**Comment by PhilterPaper** [comment]
Before I go and write a validation routine incorporating the thoughts above, I see that there are a number of data validation packages on CPAN. Can anyone recommend one or another? I would like to avoid requiring loading of a heavyweight package (unless it's already Core), and prefer Core to non-Core packages. The idea is avoid requiring yet another prerequisite package if I can help it, especially if it's non-Core. Any experiences to report?
